### PR TITLE
[ntuple] Allow for storing fields in different column representations

### PR DESCRIPTION
**Pull request by jblomer**
Add the concept of column representations to fields. Every fields can have a set of possible column representations for writing and additional column representations for reading. When reading, the column representation is fixed by the on-disk data. On writing, there is a default column representation but the user can pick a different one.

This provides the framework to store data in different encodings (e.g., split, unsplit) and to load data from "compatible columns", e.g. `kInt32` column into `std::int64` type.

As an example, the PR adds support for split and unsplit storage of `double`s. Follow-up PRs will complete the available encodings and select split encode as the default for writing.

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 1]
```diff
@@ -130,7 +156,13 @@ protected:
    std::vector<ReadCallback_t> fReadCallbacks;
    /// C++ type version cached from the descriptor after a call to `ConnectPageSource()`
    std::uint32_t fOnDiskTypeVersion = kInvalidTypeVersion;
+   /// If set, the column representation used for serialization; otherwise the default from
+   /// GetColumnRepresentations() is used
+   std::unique_ptr<std::vector<EColumnType>> fColumnRepresentative;
```
Comment (jalopezg-git) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:161]: This is always an element in the list of column representations returned by `GetColumnRepresentations()`.  Any chance that we can avoid the copy and keep only a pointer/reference to it?

File: tree/ntuple/v7/src/RField.cxx [hunk 2]
```diff
@@ -180,6 +180,23 @@ std::tuple<void **, std::int32_t *, std::int32_t *> GetRVecDataMembers(void *rve
 
 } // anonymous namespace
 
+//------------------------------------------------------------------------------
+
+std::vector<ROOT::Experimental::EColumnType>
```
Comment (jalopezg-git) [inline tree/ntuple/v7/src/RField.cxx:185]: I think we could return a `const` reference instead.

File: tree/ntuple/v7/src/RField.cxx [hunk 3]
```diff
@@ -389,37 +415,46 @@ void ROOT::Experimental::Detail::RFieldBase::Flush() const
    }
 }
 
+std::vector<ROOT::Experimental::EColumnType> ROOT::Experimental::Detail::RFieldBase::GetColumnRepresentative() const
```
Comment (jalopezg-git) [inline tree/ntuple/v7/src/RField.cxx:418]: I think we could return a `const` reference.

**Review by jalopezg-git — COMMENTED** [review]
Really nice work!  Thanks!  :slightly_smiling_face: 
I have attached some comments to the review.

**Review by jalopezg-git — APPROVED** [review]
LGTM!

File: tree/ntuple/v7/src/RColumnElement.cxx [hunk 4]
```diff
@@ -58,6 +58,9 @@ ROOT::Experimental::Detail::RColumnElementBase::Generate<void>(EColumnType type)
    return nullptr;
 }
 
+template std::unique_ptr<ROOT::Experimental::Detail::RColumnElementBase>
+ROOT::Experimental::Detail::RColumnElementBase::Generate<void>(EColumnType type);
+
```
Comment (hahnjo) [inline tree/ntuple/v7/src/RColumnElement.cxx:63]: @jalopezg-git @jblomer Clang complains about this:
```
/home/jhahnfel/ROOT/src/tree/ntuple/v7/src/RColumnElement.cxx:62:49: warning: explicit instantiation of 'Generate<void>' that occurs after an explicit specialization has no effect [-Winstantiation-after-specialization]
ROOT::Experimental::Detail::RColumnElementBase::Generate<void>(EColumnType type);
                                                ^
/home/jhahnfel/ROOT/src/tree/ntuple/v7/inc/ROOT/RColumnElement.hxx:597:57: note: previous template specialization is here
std::unique_ptr<RColumnElementBase> RColumnElementBase::Generate<void>(EColumnType type);
                                                        ^
1 warning generated.
```

I think this points to a real problem: If a TU only includes `RColumnElement.hxx` and instantiates `Generate<void>`, isn't it an ODR violation to define a different implementation here in `RColumnElement.cxx`?

Comment (jalopezg-git) [reply tree/ntuple/v7/src/RColumnElement.cxx:63]: > I think this points to a real problem: If a TU only includes `RColumnElement.hxx` and instantiates `Generate<void>`, isn't it an ODR violation to define a different implementation here in `RColumnElement.cxx`?

Actually not, the template specialization for `CppT = void` is declared in `RColumnElement.hxx`, but only defined in `RColumnElement.cxx`.  Any translation unit including `RColumnElement.hxx` should see the symbol as `extern`.

I created https://github.com/root-project/root/pull/12280; that should fix it.  :-) - my fault, anyways, as I suggested that [here](https://github.com/root-project/root/pull/12133#discussion_r1091805089).

Comment (hahnjo) [reply tree/ntuple/v7/src/RColumnElement.cxx:63]: Ah right, there's the declaration in the header, makes sense.

**Pull request merged**